components/camel-reactive-streams/src/main/java/org/apache/camel/component/reactive/streams/util/ConvertingPublisher.java

```diff
@@ -0,0 +1,113 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.camel.component.reactive.streams.util;
+
+import java.util.Objects;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import org.apache.camel.Exchange;
+import org.apache.camel.TypeConversionException;
+import org.reactivestreams.Publisher;
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A publisher that converts Camel {@code Exchange}s into the given type.
+ */
+public class ConvertingPublisher<R> implements Publisher<R> {
+
+    private static final Logger LOG = LoggerFactory.getLogger(ConvertingPublisher.class);
+
+    private Publisher<Exchange> delegate;
+
+    private Class<R> type;
+
+    public ConvertingPublisher(Publisher<Exchange> delegate, Class<R> type) {
+        Objects.requireNonNull(delegate, "delegate publisher cannot be null");
+        this.delegate = delegate;
+        this.type = type;
+    }
+
+    @Override
+    public void subscribe(Subscriber<? super R> subscriber) {
+        delegate.subscribe(new Subscriber<Exchange>() {
+
+            private AtomicBoolean active = new AtomicBoolean(true);
+
+            private Subscription subscription;
+
+            @Override
+            public void onSubscribe(Subscription subscription) {
+                if (subscription == null) {
+                    throw new NullPointerException("subscription is null");
+                }
+                if (this.subscription != null) {
```

Comment:
It's in the specs that a subscriber can have at most one active subscription and when it's given a second subscription it must cancel it immediately. The TCK has some tests about this. But you're right: if you subscribe to *the same* subscription twice it doesn't work :)